src/gui/transferlistsortmodel.cpp

```diff
@@ -199,6 +199,7 @@ int TransferListSortModel::compare(const QModelIndex &left, const QModelIndex &r
     case TransferListModel::TR_POPULARITY:
         return customCompare(leftValue.toReal(), rightValue.toReal());
 
+    case TransferListModel::TR_PRIVATE:
```

Comment:
Try to add the following function:
```c++
    int customCompare(const QVariant &left, const QVariant &right)
    {
        const bool isLeftValid = left.isValid();
        const bool isRightValid = right.isValid();

        if (isLeftValid == isRightValid)
            return threeWayCompare(left, right);
        return isLeftValid ? -1 : 1;
    }
``` 

...and use it to compare "privateness":
```c++
case TransferListModel::TR_PRIVATE:
    return customCompare(leftValue, rightValue);
```

Reply:
> Try to add the following function:
> 
> ```c++
>     int customCompare(const QVariant &left, const QVariant &right)
>     {
>         const bool isLeftValid = left.isValid();
>         const bool isRightValid = right.isValid();
> 
>         if (isLeftValid == isRightValid)
>             return threeWayCompare(left, right);
>         return isLeftValid ? -1 : 1;
>     }
> ```

Or maybe convert an existing one into a template:
```c++
// this
int customCompare(const QDateTime &left, const QDateTime &right)
// into
template <typename T>
int customCompare(const T &left, const T &right)
```

Reply:
Please forgive me for maybe being slow in understanding.

In my latest commit, I simply moved the `case TransferListModel::TR_PRIVATE:` from a `threeWayCompare` to a `customCompare`. 

However, I did not convert an existing function into a template, because I believe that is already existing?:

https://github.com/qbittorrent/qBittorrent/blob/698f8dcb9cacf36a1e42ce7abfa7ac3d80df6476/src/gui/transferlistsortmodel.cpp#L72-L85

Reply:
> However, I did not convert an existing function into a template, because I believe that is already existing?:

Oh, right, I overlooked it (it's not very convenient to do all this through a smartphone browser).
Unfortunately, it is limited only to arithmetic types.
But it's not a problem to make a more generic template. You can delete all instances of `customCompare` except the one that accept `TagSet` as parameters, and add the following code instead:
```c++
    // consider negative values as invalid
    template <typename T>
        requires std::is_arithmetic_v<T>
    bool isValid(const T value)
    {
        return (value >= 0);
    }

    template <typename T>
    bool isValid(const T &value)
    {
        return value.isValid();
    }

    template <typename T>
    int customCompare(const T &left, const T &right)
    {
        const bool isLeftValid = isValid(left);
        const bool isRightValid = isValid(right);

        if (isLeftValid && isRightValid)
            return threeWayCompare(left, right);
        if (!isLeftValid && !isRightValid)
            return 0;
        return isLeftValid ? -1 : 1;
    }
```

Reply:
Done - had to add a small refinement, since the 2nd template was also triggered for arithmetic types and the CI complained about template overloading. Hope the refinement makes sense?

[d486622](https://github.com/qbittorrent/qBittorrent/pull/20951/commits/d486622ad8a104657cb24aabc6226d272d7b41a6)

Reply:
> CI complained about template overloading.

Hmm... Clang compiled this without a problem.
`std::enable_if_t` looks cumbersome. I'd rather use a more modern approach with `if constexpr`:
```c++
    template <typename T>
    bool isValid(const T &value)
    {
        if constexpr (std::is_arithmetic_v<T>)
            return (value >= 0); // consider negative values as invalid
        else
            return value.isValid();
    }
```

Reply:
...or better add requirement to second template:
```c++
    template <typename T>
    concept Validateable = requires (T t) { {t.isValid()} -> std::same_as<bool>; };

    template <Validateable T>
    bool isValid(const T &value)
    {
        return value.isValid();
    }

    // consider negative values as invalid
    template <typename T>
        requires std::is_arithmetic_v<T>
    bool isValid(const T value)
    {
        return (value >= 0);
    }
```

Reply:
updated ✔
https://github.com/qbittorrent/qBittorrent/pull/20951/commits/05c4f171d553e40dcdddf4b6db120531476134e4

Reply:
done - https://github.com/qbittorrent/qBittorrent/pull/20951/commits/d8c743994449792b2ab1451ebc5a09d08ddb4ec4